<a href="https://colab.research.google.com/github/jeperniag/BussinessIntelligence/blob/main/Reto1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Cargar librerias

In [ ]:
import pandas as pd
import numpy as np

Cargar base de datos
1. ¿Cuál es el número de datos total de la base de datos?

In [ ]:
nxl='/content/1. BD1_Longitudinal Survey_Int.xlsx'
XDB=pd.read_excel(nxl,sheet_name=0)

num_filas = XDB.shape[0]  # Obtiene el número de filas
num_columnas = XDB.shape[1]  # Obtiene el número de columnas

print("Número de filas:", num_filas)
print("Número de columnas:", num_columnas)

Número de filas: 12686
Número de columnas: 24


In [ ]:
#Se seleccionan las variables de interés
XDB=XDB[['Age','Urban','Mother_Edu', 'Father_Edu','FamilySize','Self_Esteem','Height','Weight','Income']]
XDB.head(10)

#Se separan las variables de entrada y salida
XD=XDB[['Age','Mother_Edu', 'Father_Edu','FamilySize','Self_Esteem','Height','Weight','Income']] #Variables de Entrada
yd=XDB[['Urban']] #Variable de Salida - Variable Pronóstico - Deseada

# Eliminar las filas con valores NaN en 'yd' antes de entrenar el modelo
# Esto asegura que 'yd' no contenga ningún valor NaN

XD=XDB[['Age','Mother_Edu', 'Father_Edu','FamilySize','Self_Esteem','Height','Weight','Income']] #Variables de Entrada
yd=XDB[['Urban']] #Variable de Salida - Variable Pronóstico - Deseada

XDB=XDB.dropna() #Evitar celdas vacias

num_filas = XDB.shape[0]  # Obtiene el número de filas
num_columnas = XDB.shape[1]  # Obtiene el número de columnas

print("Número de filas:", num_filas)
print("Número de columnas:", num_columnas)

Número de filas: 5948
Número de columnas: 9


•	Determinar la categoría a la que pertenece una persona con las siguientes características mediante la utilización del modelo Naive Bayes y Clustering: [5000000,37,82,245,45,17.8,0.5,1.4,0.4]







In [ ]:
from sklearn.naive_bayes import GaussianNB #from antes de import es prestamo interbibliotecario

mnb=GaussianNB() #Aloja el modelo - Campana
mnb.fit(XD,yd) #Entrena el modelo - Sacar media, de, valores de pertenencia

#Obtenemos la información del modelo
u=mnb.theta_ #Obtenemos la media
print("La media es:\n",u)
var=mnb.var_ #Obtenemos la varianzas - O2 es la varianza y O es la desviación
sigma=np.sqrt(var) #Se calcula la desviación estandar
print("La desviación es:\n",sigma)

#Se determinan los limites superiores e inferiores
LS=u+sigma
LI=u-sigma
print("El limite superior es:\n",LS)
print("El limite inferior es:\n",LI)

#Evaluamos un solicitantes
XDS=[5000000,37,82,245,45,17.8,0.5,1.4,0.4] #Como un valor de entrada -
ydp=mnb.predict(XDS) #Se obtiene la predicción
print("La predicción de la persona es:\n",ydp)

if ydp==0:
  print("La persona vive en un área rural")

if ydp==1:
  print("La persona vive en un área urbana")

La media es:
 [[1.83640032e+01 1.07312349e+01 1.02711864e+01 4.73930589e+00
  2.19951574e+01 6.72429379e+01 1.49003228e+02 3.01248854e+04]
 [1.83618603e+01 1.11613931e+01 1.12331705e+01 4.68103631e+00
  2.25744319e+01 6.69116585e+01 1.44750903e+02 3.32736742e+04]]
La desviación es:
 [[2.47260545e+00 3.15661127e+00 3.88837565e+00 2.55226748e+00
  4.21650807e+00 4.08228636e+00 3.03486350e+01 3.16482374e+04]
 [2.47344359e+00 3.40066236e+00 4.11549881e+00 2.44205851e+00
  4.19763081e+00 4.20193102e+00 3.00285301e+01 3.60897302e+04]]
El limite superior es:
 [[2.08366087e+01 1.38878461e+01 1.41595621e+01 7.29157338e+00
  2.62116655e+01 7.13252242e+01 1.79351863e+02 6.17731228e+04]
 [2.08353039e+01 1.45620554e+01 1.53486693e+01 7.12309482e+00
  2.67720628e+01 7.11135895e+01 1.74779433e+02 6.93634044e+04]]
El limite inferior es:
 [[ 1.58913978e+01  7.57462359e+00  6.38281079e+00  2.18703841e+00
   1.77786493e+01  6.31606515e+01  1.18654593e+02 -1.52335198e+03]
 [ 1.58884167e+01  7.76073072e+00

/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


2. ¿Cuál es el porcentaje de los datos que hacen parte de cada una de las categorías de clasificación?

In [ ]:
# Porcentaje de cada categoría en "Urban"
categories = XDB['Urban'].value_counts(normalize=True) * 100
print('Porcentaje de datos en cada categoría:\n', categories)
#1.0 → 79.16%: Significa que el 79.16% de las personas en la base de datos viven en áreas urbanas.
#0.0 → 20.83%: Indica que el 20.83% de las personas viven en áreas rurales.

Porcentaje de datos en cada categoría:
 Urban
1.0    79.169469
0.0    20.830531
Name: proportion, dtype: float64


3. Determinar el intervalo en el cual se encuentra la variación de la Edad

4. Determinar el intervalo en el cual se encuentra la variación de la Altura

5. Determinar el intervalo en el cual se encuentra la variación del Peso de las Personas

In [ ]:
#Intervalos de variación de Edad, Altura y Peso
print('Rango de Edad', (XDB['Age'].min(), XDB['Age'].max()))
print('Rango de Altura', (XDB['Height'].min(), XDB['Height'].max()))
print('Rango de Peso', (XDB['Weight'].min(), XDB['Weight'].max()))


Rango de Edad (15, 22)
Rango de Altura (51.0, 80.0)
Rango de Peso (72.0, 375.0)


Evaluar el desempeño del modelo

6. ¿Cuál es la exactitud alcanzada por el modelo?

7. ¿Cuál es la tasa de error alcanzada por el modelo?

8. ¿Cuál es la Sensibilidad alcanzada por el modelo?

9. ¿Cuál es la Especificidad alcanzada por el modelo?

In [ ]:
from sklearn.metrics import confusion_matrix

#Haga un barrido de los datos
vdp2=mnb.predict(XD) #Haga el pronóstico para los datos base

cm=confusion_matrix(yd,vdp2) #Cree la matriz de confusión
print("La matriz de confusión es:\n",cm)
VN=cm[0,0];FP=cm[0,1];FN=cm[1,0];VP=cm[1,1]

#Metricas Desempeño
Ex=(VP+VN)/(VP+VN+FP+FN) #Exactitud : Comportamiento general urbano
print("La exactitud del modelo es:\n",Ex)

TE=(FN+FP)/(VP+VN+FP+FN) #Tasa de Error
print("El error del modelo es:\n",TE)

Se=VP/(VP+FN) #Sensibilidad : Comportamiento correcto urbano
              #Sobre el total de los urbanos
print("La sensibilidad del modelo es:\n",Se)

Sp=VN/(VN+FP) #Especificidad : Comportamiento correcto no urbano
              #Sobre el total de los no urbanos (rurales)
print("La especificidad del modelo es:\n",Sp)

Ur=VP/(VP+FP) #Precisión : Comportamiento correcto urbano
             #Sobre el total del FP + VP pronóstico
print("La precisión del modelo es:\n",Ur)

Ru=VN/(VN+FN) #Precisión Negativa : Comportamiento correcto no urbano
               #Sobre el total del FN + VN pronóstico
print("La precisión negativa del modelo es:\n",Ru)

NameError: name 'mnb' is not defined


**Análisis de Resultados**
 De acuerdo con la base de datos suministrada, se tomaron como referencia para saber si la persona vive en el área urbana o rural las variables socioeconómicas como Edad, Educación de los Padres, Estatura, Peso, Ingresos, Escala de Autoestima y Tamaño de la Familia. La base de datos posee un total de 12686 datos, divididos en 79,1% datos de personas que viven en el área urbana y un total de 20,8% de personas que viven en el área rural. De manera general, el modelo alcanzó una exactitud cercana al 80% en promedio (79,1%), lo que demuestra un comportamiento muy aceptable frente a las personas que viven en el área urbana. Sin embargo, es relevante destacar, que el modelo tuvo una especificidad muy baja (0,024%), lo que muestra un mal comportamiento frente a las personas que viven en la zona rural. No obstante, al analizar el índice de precisión, el cual alcanzó un valor cercano al 80% en promedio (79%), muestra que a pesar de todo el modelo tiene un buen criterio frente a la toma de decisiones en las personas que viven en el área urbana. Por lo que a modo general, se puede decir que tiene el modelo tiene un alto desempeño en la identificación de áreas urbanas además, de tener una sensibilidad demasiada alta del casi 100% lo que beneficia al modelo.